frontend/app/src/util/AppNavigation.ts

```diff
@@ -140,6 +149,128 @@ export class V1Strategy {
       ) ?? this.appPages[0]
     )
   }
+
+  clearPageElements(
+    _elements: AppRoot,
+    mainScriptHash: string,
+    sidebarElements: BlockNode | undefined
+  ): AppRoot {
+    return AppRoot.empty(mainScriptHash, false, sidebarElements)
+  }
+}
+
+export class V2Strategy {
+  readonly parent: AppNavigation
+
+  mainScriptHash: string | null
+
+  appPages: IAppPage[]
+
+  mainPage: IAppPage | null
+
+  hideSidebarNav: boolean | null
+
+  constructor(parent: AppNavigation) {
+    this.parent = parent
+    this.mainScriptHash = null
+    this.appPages = []
+    this.mainPage = null
+    this.hideSidebarNav = null
+  }
+
+  handleNewSession(newSession: NewSession): MaybeStateUpdate {
+    this.mainScriptHash = newSession.mainScriptHash
+    this.hideSidebarNav = newSession.config?.hideSidebarNav ?? false
+
+    // We do not know the page name, so use an empty string version
+    document.title = " · Streamlit"
+
+    return [{ hideSidebarNav: this.hideSidebarNav }, () => {}]
+  }
+
+  handlePagesChanged(_pagesChangedMsg: PagesChanged): MaybeStateUpdate {
+    // This message does not apply to V2
+    return undefined
+  }
+
+  handlePageNotFound(pageNotFound: PageNotFound): MaybeStateUpdate {
+    const { pageName } = pageNotFound
+    this.parent.onPageNotFound(pageName)
+
+    return [
+      { currentPageScriptHash: this.mainScriptHash ?? "" },
+      () => {
+        this.parent.hostCommunicationMgr.sendMessageToHost({
+          type: "SET_CURRENT_PAGE_NAME",
+          currentPageName: "",
+          currentPageScriptHash: this.mainScriptHash ?? "",
+        })
+      },
+    ]
+  }
+
+  handleNavigation(navigationMsg: Navigation): MaybeStateUpdate {
+    const { sections, position, appPages } = navigationMsg
```

Comment:
super extreme nit: can we try to sort these variables alphabetically? 😅 it helps human to scan a list; this case is extreme as we only have 3 variables, but we have other places where we started to clean it up a little bit